src/Microsoft.VisualStudio.LanguageServices.Razor/DefaultVisualStudioOpenDocumentManager.cs

```diff
@@ -0,0 +1,167 @@
+using System.Linq;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Razor;
+using Microsoft.VisualStudio.Editor;
+using Microsoft.VisualStudio.Editor.Razor;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+
+namespace Microsoft.VisualStudio.LanguageServices.Razor
+{
+    [Export(typeof(VisualStudioOpenDocumentManager))]
+    internal class DefaultVisualStudioOpenDocumentManager : VisualStudioOpenDocumentManager
+    {
+        private readonly IVsFileChangeEx _fileChangeService;
+        private readonly IVsRunningDocumentTable _runningDocumentTable;
+        private readonly IVsEditorAdaptersFactoryService _editorAdaptersFactoryService;
+        private readonly ForegroundDispatcher _foregroundDispatcher;
+        private readonly ErrorReporter _errorReporter;
+
+        private List<VisualStudioDocumentTracker> _documents;
+        private Dictionary<string, ViewImportChangeTracker> _viewImportChangeTrackerCache;
+
+        [ImportingConstructor]
+        public DefaultVisualStudioOpenDocumentManager(
+            [Import(typeof(SVsServiceProvider))] IServiceProvider serviceProvider,
+            [Import(typeof(VisualStudioWorkspace))] Workspace workspace,
+            IVsEditorAdaptersFactoryService editorAdaptersFactoryService)
+        {
+            if (serviceProvider == null)
+            {
+                throw new ArgumentNullException(nameof(serviceProvider));
+            }
+
+            if (workspace == null)
+            {
+                throw new ArgumentNullException(nameof(workspace));
+            }
+
+            if (editorAdaptersFactoryService == null)
+            {
+                throw new ArgumentNullException(nameof(editorAdaptersFactoryService));
+            }
+
+            _fileChangeService = serviceProvider.GetService(typeof(SVsFileChangeEx)) as IVsFileChangeEx;
+            _runningDocumentTable = serviceProvider.GetService(typeof(SVsRunningDocumentTable)) as IVsRunningDocumentTable;
+            _editorAdaptersFactoryService = editorAdaptersFactoryService;
+
+            _documents = new List<VisualStudioDocumentTracker>();
+            _viewImportChangeTrackerCache = new Dictionary<string, ViewImportChangeTracker>(StringComparer.OrdinalIgnoreCase);
+
+            _foregroundDispatcher = workspace.Services.GetRequiredService<ForegroundDispatcher>();
+            _errorReporter = workspace.Services.GetRequiredService<ErrorReporter>();
+        }
+
+        public override IReadOnlyList<VisualStudioDocumentTracker> OpenDocuments => _documents;
+
+        public override void AddDocument(VisualStudioDocumentTracker tracker)
+        {
+            _foregroundDispatcher.AssertForegroundThread();
+
+            if (!_documents.Contains(tracker))
+            {
+                _documents.Add(tracker);
+
+                if (!tracker.TextBuffer.Properties.TryGetProperty<VisualStudioRazorParser>(typeof(VisualStudioRazorParser), out var parser))
```

Comment:
Couldn't use `RazorEditorFactoryService` to get the parser here because #circularimport